From 8c2038e90cccfb512910154ac9975bc897f2afe0 Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Sun, 19 Oct 2014 17:10:49 +0200
Subject: [PATCH] video: sunxi hdmi: Fix switching to interlaced modes.

Corrects timing calculation and output switching for interlaced video modes.

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 drivers/video/sunxi/disp/disp_hdmi.c |   17 +++++++++++++++--
 drivers/video/sunxi/disp/disp_lcd.c  |    6 +++++-
 2 files changed, 20 insertions(+), 3 deletions(-)

diff --git a/drivers/video/sunxi/disp/disp_hdmi.c b/drivers/video/sunxi/disp/disp_hdmi.c
index 1043215..e77a2d2 100644
--- a/drivers/video/sunxi/disp/disp_hdmi.c
+++ b/drivers/video/sunxi/disp/disp_hdmi.c
@@ -226,9 +226,18 @@ void videomode_to_video_timing(struct __disp_video_timing *video_timing,
 	video_timing->VBP = mode->upper_margin + mode->vsync_len;
 	video_timing->VFP = mode->lower_margin;
 	video_timing->VPSW = mode->vsync_len;
-	if (mode->vmode & FB_VMODE_INTERLACED)
+
+	if (mode->vmode & FB_VMODE_INTERLACED) {
 		video_timing->I = true;
 
+		video_timing->VBP /= 2;
+		video_timing->VFP /= 2;
+		video_timing->VPSW /= 2;
+
+		if ((video_timing->HT * (video_timing->VT + 1) * (mode->refresh/2)) == video_timing->PCLK)
+			video_timing->VT++;
+	}
+
 	if (mode->sync & FB_SYNC_HOR_HIGH_ACT)
 		video_timing->HSYNC = true;
 
@@ -272,7 +281,11 @@ __s32 BSP_disp_set_videomode(__u32 sel, const struct fb_videomode *mode)
 		goto failure;
 
 	gdisp.screen[sel].hdmi_mode = DISP_TV_MODE_EDID;
-	gdisp.screen[sel].b_out_interlace = new_video_timing->I;
+
+	if (gdisp.screen[sel].b_out_interlace != new_video_timing->I) {
+	      gdisp.screen[sel].b_out_interlace = new_video_timing->I;
+	      Disp_set_out_interlace(sel);
+	}
 
 	kfree(old_video_timing);
 	kfree(new_video_timing);
diff --git a/drivers/video/sunxi/disp/disp_lcd.c b/drivers/video/sunxi/disp/disp_lcd.c
index 1596835..614201a 100644
--- a/drivers/video/sunxi/disp/disp_lcd.c
+++ b/drivers/video/sunxi/disp/disp_lcd.c
@@ -1840,8 +1840,12 @@ __s32 BSP_disp_get_videomode(__u32 sel, struct fb_videomode *videomode)
 	videomode->xres = BSP_disp_get_screen_width(sel);
 	videomode->yres = BSP_disp_get_screen_height(sel);
 
-	if (interlaced)
+	if (interlaced) {
 		videomode->vmode = FB_VMODE_INTERLACED;
+		videomode->lower_margin *= 2;
+		videomode->upper_margin *= 2;
+		videomode->vsync_len *= 2;
+	}
 
 	if (vsync)
 		videomode->sync = FB_SYNC_VERT_HIGH_ACT;
-- 
1.7.9.5

